ovis_robotiq_gripper/nodes/ovis_gripper_control.py

```diff
@@ -0,0 +1,131 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+# Copyright (c) 2012, Robotiq, Inc.
+# Revision $Id$
+
+# Base imports
+import sys
+import os
+
+# Message imports
+from ovis_robotiq_gripper.msg import _OvisGripperPosition as positionMsg
+from ovis_robotiq_gripper.msg import _OvisGripper_robot_output as outputMsg
+from ovis_robotiq_gripper.msg import _OvisGripper_robot_input as inputMsg
+import gripper_constants as cons
+
+# Custom, Robotiq and ROS imports
+import robotiq_modbus_rtu.comModbusRtu
+import ovis_robotiq_gripper.baseRobotiq2FGripper
+import rospy
+import roslib
+roslib.load_manifest('robotiq_2f_gripper_control')
+roslib.load_manifest('robotiq_modbus_rtu')
+
+# Global declararions
+command = outputMsg.OvisGripper_robot_output()
+current_position = inputMsg.OvisGripper_robot_input()
+
+
+def command_callback(input):
+    global command
+    global current_position
+
+    # Default values setup
+    command.rACT = cons.ON  # activates the gripper
+    command.rGTO = cons.ON  # Go to requested position
+    command.rFR = cons.MAX_FORCE  # maximum force
+    command.rSP = cons.MAX_SPEED
+
+    if input.position == 0:
+        command.rACT = cons.OFF  # activates the gripper
+        command.rGTO = cons.OFF  # Go to requested position
+        rospy.logdebug("Debug info: position = 2 and command sent is OFF")
+
+    if input.position == 1:
+        rospy.logdebug("Debug info: position = 1 and command sent is ON")
+
+    elif input.position == 2 and current_position.gPO <= 3:
+        command.rPR = cons.CLOSE  # close
+
+    elif input.position == 2:
+        command.rPR = cons.OPEN  # open
+
+def mainLoop(device):
+
+    # Gripper is a 2F with a TCP connection
+    gripper = ovis_robotiq_gripper.baseRobotiq2FGripper.robotiqbaseRobotiq2FGripper()
+    gripper.client = robotiq_modbus_rtu.comModbusRtu.communication()
+
+    # We connect to the address received as an argument
+    gripper.client.connectToDevice(device)
+
+    rospy.init_node("ovis_robotiq_gripper")
+
+    # The Gripper status is published on the topic named 'gripper_input'
+    pubInput = rospy.Publisher(
+        'gripper_input', inputMsg.OvisGripper_robot_input, queue_size=1)
+
+    # The Gripper command is published on the topic named 'gripper_output'
+    pubOutput = rospy.Publisher(
+        'gripper_output', outputMsg.OvisGripper_robot_output, queue_size=1)
+
+    #boot_sequence(gripper, pubOutput)
+
+    # We loop
+    while not rospy.is_shutdown():
+        global command
+        global current_position
+
+        # Get and publish the Gripper status
+        status = gripper.getStatus()
+        pubInput.publish(status)
+        current_position = status
+
+        # Get the most recent command
+        rospy.Subscriber("/ovis/gripper/position_goal",
```

Comment:
I'm not necessary familiar with ROS Python, is that the idiomatic way of doing it in Python? If I read it correctly, every time we loop we subscribe to a topic, couldn't this be done before, outside the while loop?

Reply:
If I understand correctly, the standard way for a publisher is the while() loop and for a subscriber its rospy.spin().
In a combined node it becomes blurry. The end point of both are functionally the same. 
To use rospy.spin(), I would have to put the publishers in the subscriber callbacks and they would be conditional to said callback, making debugging less convenient. 

Sources: http://wiki.ros.org/rospy/Overview/Publishers%20and%20Subscribers and https://get-help.robotigniteacademy.com/t/what-is-difference-between-rospy-spin-and-while-not-rospy-is-shutdown/1849/5

Reply:
> The end point of both are functionally the same.

Humm I'm not convince, publishing see publish data to a topic to a certain rate. Subscriber wait to receive a published topic. The Subscriber doesn't require a rate like the publisher. So having it inside the loop doesn't make sense to me. 

Can you try and see what happen if you put it outside the loop? This will test my theory, if you have communication issues with the subscriber we keep it inside the `main_loop` and if it doesn't affect it, we'll keep it outside the loop.

Also, much like the problem we had with the E-Stop node where it would take a lot of CPU, we ought to add a rate also for this node.